File: Loenn/entities/grappleController.lua
local grappleController = {}

grappleController.name = "Casju0TrailMix/GrappleController"
grappleController.depth = 0
grappleController.placements = {
   {
      name = "grapple_controller_lani",
      data = {
         enabled = true,
         inventoryType = "AlwaysEquipped",
         controlType = "ReplaceGrab",
         harpoonMode = false,
         canShootWhenTired = false,
         addMomentumToShot = true,
         playerReleaseMomentum = "RetainGrappleSpeed",
         playerReleaseMomentumSpeedCap = -1,
         playerReleaseSpeedBoost = 40,
         freezeFrameLength = "Long",
         contactPauseLength = "None",
         shockwave = "Small",
         stringColor = "FFFFFF",
         contactColor = "FFB6C1",
         reelColor = "ADD8E6",
         canWallCancel = true,
         canItemCancel = true,
         canWallJumpCancel = true,
         canItemJumpCancel = true,
         jumpCancelStaminaCost = 27.5,
         canWallDashCancel = true,
         canItemDashCancel = true,
         pufferBehavior = "PullPuffer",
         yOffset = -2,
         shootSpeed = 360,
         retractSpeed = 480,
         wallPullSpeed = 240,
         itemPullSpeed = 240,
         minShootDuration = 0.1,
         maxShootDuration = 0.25,
         cooldown = 0,
      },
   },
   {
      name = "grapple_controller_hornet",
      data = {
         enabled = true,
         inventoryType = "AlwaysEquipped",
         controlType = "UserBinding",
         harpoonMode = true,
         canShootWhenTired = true,
         addMomentumToShot = false,
         playerReleaseMomentum = "RetainGrappleSpeed",
         playerReleaseMomentumSpeedCap = 90,
         playerReleaseSpeedBoost = 0,
         freezeFrameLength = "None",
         contactPauseLength = "Short",
         shockwave = "Small",
         stringColor = "FFFFFF",
         contactColor = "FFB6C1",
         reelColor = "ADD8E6",
         canWallCancel = false,
         canItemCancel = false,
         canWallJumpCancel = false,
         canItemJumpCancel = false,
         jumpCancelStaminaCost = 27.5,
         canWallDashCancel = false,
         canItemDashCancel = false,
         pufferBehavior = "PullPlayer",
         yOffset = -2,
         shootSpeed = 600,
         retractSpeed = 600,
         wallPullSpeed = 600,
         itemPullSpeed = 600,
         minShootDuration = 0.1,
         maxShootDuration = 0.1,
         cooldown = 1,
      },
   },
}

grappleController.fieldInformation = {
   inventoryType = {
      options = {
         "RequiresRefill",
         "AlwaysEquipped",
      },
      editable = false,
   },
   controlType = {
      options = {
         "ReplaceGrab",
         "ReplaceDash",
         "UserBinding",
      },
      editable = false,
   },
   playerReleaseMomentum = {
      options = {
         "RetainGrappleSpeed",
         "RevertSpeed",
      },
      editable = false,
   },
   freezeFrameLength = {
      options = {
         "None",
         "Short",
         "Long",
         "VeryLong",
      },
      editable = false,
   },
   contactPauseLength = {
      options = {
         "None",
         "Short",
         "Long",
         "VeryLong",
      },
      editable = false,
   },
   shockwave = {
      options = {
         "None",
         "Small",
         "Large",
      },
      editable = false,
   },
   stringColor = {
      fieldType = "color",
   },
   contactColor = {
      fieldType = "color",
   },
   reelColor = {
      fieldType = "color",
   },
   pufferBehavior = {
      options = {
         "Ignore",
         "PullPuffer",
         "PullPlayer",
      },
      editable = false,
   },
   yOffset = {
      fieldType = "integer",
      minimumValue = -5,
      maximumValue = 5,
   },
}

grappleController.canResize = { false, false }
grappleController.texture = "objects/Casju0TrailMix/loenn/grappleController"

return grappleController

File: Source/Entities/Controllers/GrappleController.cs
namespace Celeste.Mod.Casju0TrailMix.Entities;

[Tracked]
[CustomEntity("Casju0TrailMix/GrappleController")]
public class GrappleController : Entity
{
    public const string flag = "Casju0TrailMix/grappleEnabled";
    static GrappleSubMenu Settings { get => Casju0TrailMixModule.Settings.GrappleSettings; }

    bool enabled;

    public GrappleController(EntityData data, Vector2 position)
    {
        enabled = data.Bool(flag, true);
        Settings.InventoryType = data.Enum<GrappleSubMenu.InventoryTypes>("inventoryType", GrappleSubMenu.InventoryTypes.AlwaysEquipped);
        Settings.ControlType = data.Enum<GrappleSubMenu.ControlTypes>("controlType", GrappleSubMenu.ControlTypes.ReplaceGrab);
        Settings.HarpoonMode = data.Bool("harpoonMode", false);
        Settings.CanShootWhenTired = data.Bool("canShootWhenTired", false);
        Settings.AddMomentumToShot = data.Bool("addMomentumToShot", true);
        Settings.PlayerReleaseMomentum = data.Enum<GrappleSubMenu.PlayerReleaseMomentums>("playerReleaseMomentum", GrappleSubMenu.PlayerReleaseMomentums.RetainGrappleSpeed);
        Settings.PlayerReleaseMomentumSpeedCap = data.Float("playerReleaseMomentumSpeedCap", -1f);
        Settings.PlayerReleaseSpeedBoost = data.Float("playerReleaseSpeedBoost", 40f);
        Settings.FreezeFrameLength = data.Enum<GrappleSubMenu.FreezeFrameLengths>("freezeFrameLength", GrappleSubMenu.FreezeFrameLengths.Long);
        Settings.ContactPauseLength = data.Enum<GrappleSubMenu.FreezeFrameLengths>("contactPauseLength", GrappleSubMenu.FreezeFrameLengths.None);
        Settings.Shockwave = data.Enum<GrappleSubMenu.Shockwaves>("shockwave", GrappleSubMenu.Shockwaves.Small);
        Settings.StringColor = data.HexColor("stringColor", Calc.HexToColor("FFFFFF"));
        Settings.ContactColor = data.HexColor("contactColor", Calc.HexToColor("FFB6C1"));
        Settings.ReelColor = data.HexColor("reelColor", Calc.HexToColor("ADD8E6"));
        Settings.CanWallCancel = data.Bool("canWallCancel", true);
        Settings.CanItemCancel = data.Bool("canItemCancel", true);
        Settings.CanWallJumpCancel = data.Bool("canWallJumpCancel", true);
        Settings.CanItemJumpCancel = data.Bool("canItemJumpCancel", true);
        Settings.JumpCancelStaminaCost = data.Float("jumpCancelStaminaCost", 27.5f);
        Settings.CanWallDashCancel = data.Bool("canWallDashCancel", true);
        Settings.CanItemDashCancel = data.Bool("canItemDashCancel", true);
        Settings.PufferBehavior = data.Enum<GrappleSubMenu.PufferBehaviors>("pufferBehavior", GrappleSubMenu.PufferBehaviors.PullPuffer);
        Settings.YOffset = data.Int("yOffset", -2);
        Settings.ShootSpeed = data.Float("shootSpeed", 360f);
        Settings.RetractSpeed = data.Float("retractSpeed", 480f);
        Settings.WallPullSpeed = data.Float("wallPullSpeed", 240f);
        Settings.ItemPullSpeed = data.Float("itemPullSpeed", 240f);
        Settings.MinShootDuration = data.Float("minShootDuration", 0.1f);
        Settings.MaxShootDuration = data.Float("maxShootDuration", 0.25f);
        Settings.Cooldown = data.Float("cooldown", 0f);
    }

    public override void Added(Scene scene)
    {
        (scene as Level).Session.SetFlag(flag, enabled);
    }
}
